ntoskrnl/mm/ARM3/sysldr.c

```diff
@@ -268,28 +256,168 @@ MiLocateExportName(IN PVOID DllBase,
+ * - STATUS_INVALID_PARAMETER if the image is invalid or does not
+ *   contain an Export Directory;
+ * - STATUS_ORDINAL_NOT_FOUND if the export could not be found;
+ * - STATUS_PROCEDURE_NOT_FOUND if the export is a forwarder.
+ *
+ * @remark
+ * The function does not handle forwarders, and will fail with
+ * STATUS_PROCEDURE_NOT_FOUND if the export is as such.
+ *
+ * @note
+ * See RtlFindExportedRoutineByName() below for more remarks.
+ * Used by psmgr.c LookupEntryPoint() as well.
+ **/
+NTSTATUS
+NTAPI
+RtlpFindExportedRoutineByName(
+    _In_ PVOID DllBase,
+    _In_ PCSTR ExportName,
+    _Out_ PVOID* Function,
+    _In_opt_ BOOLEAN CheckIfForwarder)
+{
+    PIMAGE_EXPORT_DIRECTORY ExportDirectory;
+    PULONG NameTable;
+    PUSHORT OrdinalTable;
+    ULONG ExportSize;
+    USHORT Ordinal;
+    PULONG ExportTable;
+    PVOID FunctionAddress;
+
+    PAGED_CODE();
+
+    /* Get the export directory */
+    ExportDirectory = RtlImageDirectoryEntryToData(DllBase,
+                                                   TRUE,
+                                                   IMAGE_DIRECTORY_ENTRY_EXPORT,
+                                                   &ExportSize);
+    if (!ExportDirectory)
+        return STATUS_INVALID_PARAMETER;
+
+    /* Setup name tables */
+    NameTable = (PULONG)((ULONG_PTR)DllBase +
+                         ExportDirectory->AddressOfNames);
+    OrdinalTable = (PUSHORT)((ULONG_PTR)DllBase +
+                             ExportDirectory->AddressOfNameOrdinals);
+
+    /* Get the ordinal */
+    Ordinal = NameToOrdinal(ExportName,
+                            DllBase,
+                            ExportDirectory->NumberOfNames,
+                            NameTable,
+                            OrdinalTable);
+
+    /* Check if we couldn't find it */
+    if (Ordinal == -1)
+        return STATUS_ORDINAL_NOT_FOUND;
+
+    /* Validate the ordinal */
+    if (Ordinal >= ExportDirectory->NumberOfFunctions)
+    {
+        /* It's not, fail */
+        return STATUS_ORDINAL_NOT_FOUND; // STATUS_PROCEDURE_NOT_FOUND;
+    }
 
     /* Resolve the address and write it */
     ExportTable = (PULONG)((ULONG_PTR)DllBase +
                            ExportDirectory->AddressOfFunctions);
-    Function = (PVOID)((ULONG_PTR)DllBase + ExportTable[Ordinal]);
+    FunctionAddress = (PVOID)((ULONG_PTR)DllBase + ExportTable[Ordinal]);
 
-    /* Check if the function is actually a forwarder */
-    if (((ULONG_PTR)Function > (ULONG_PTR)ExportDirectory) &&
-        ((ULONG_PTR)Function < ((ULONG_PTR)ExportDirectory + ExportSize)))
+    if (CheckIfForwarder)
     {
-        /* It is, fail */
-        return NULL;
+        /* Check if the function is actually a forwarder */
+        if (((ULONG_PTR)FunctionAddress > (ULONG_PTR)ExportDirectory) &&
+            ((ULONG_PTR)FunctionAddress < ((ULONG_PTR)ExportDirectory + ExportSize)))
+        {
+            ASSERT((FunctionAddress <= (PVOID)ExportDirectory) ||
```

Comment:
This should probably be `<`.

Reply:
The assert is the negation of the if-test (in order to avoid having an `ASSERT(FALSE);` instead). See discussion with Mark above.